py/src/call.py

```diff
@@ -455,15 +511,63 @@ async def do_call(
         ffc, state_selector=state_selector
     )
 
+    # using carried state as child_state is only required for fee estimation
+    # but doesn't seem to matter for regular call.
+    carried_state = carried_state.create_child_state_for_querying()
+
     state = StarknetState(state=carried_state, general_config=general_config)
     max_fee = 0
 
     output = await state.invoke_raw(
         contract_address, selector, calldata, caller_address, max_fee, signature
     )
 
-    # this is everything we need, at least so far for the "call".
-    return output.call_info.retdata
+    # return both of these as carried state is needed for the fee estimation afterwards
+    return (output, carried_state)
+
+
+def estimate_fee_after_call(call_info, carried_state):
+    from starkware.starknet.definitions.general_config import (
+        StarknetGeneralConfig,
+        N_STEPS_RESOURCE,
+    )
+    from starkware.cairo.lang.builtins.all_builtins import (
+        PEDERSEN_BUILTIN,
+        RANGE_CHECK_BUILTIN,
+        ECDSA_BUILTIN,
+        BITWISE_BUILTIN,
+        OUTPUT_BUILTIN,
+        EC_OP_BUILTIN,
+    )
+    from starkware.starknet.business_logic.transaction_fee import calculate_tx_fee
+    from starkware.starknet.business_logic.utils import get_invoke_tx_total_resources
+
+    general_config = StarknetGeneralConfig()
+
+    # given on 2022-06-07
+    general_config.cairo_resource_fee_weights.update(
```

Comment:
Do we need a way of automatically making sure we're always on par with the weights from SW? How would we know we are not in sync if those are changed? Can they change without us knowing?
Should we for example compare the estimated fee that our local machinery outputs with what the SW api gives?

Reply:
I was going to test today with transactions we have in the 230k block database but if these change, we are only going to learn from reports or upstream. They do not want to put these into cairo-lang.